DOCS-915: Document entry point file for modules and improve modules documentation

sguequierre opened the pull request:
WIP improving wording & page structure overall in light of resource registry

- separate out "create" and configure" sections to avoid overwhelming users that will be focusing on learning one part at a time: often, users will just need to see "configure" as they will be using pre-built modules.
- separate out key concepts section for the important api/model namespacing info-- might want to expand in there, and also could link from other parts of docs
- RE. docs ticket: attempting make "entry point file" clear where it is described on the page as the main program for the module, improve module wording overall

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,85 @@
+no_list: true
+---
+
+`viam-server` [manages](#management) modular {{< glossary_tooltip term_id="resource" text="resources" >}} configured on your robot like resources that are already built into the Robot Development Kit [(RDK)](/internals/rdk/).
```
```suggestion
`viam-server` [manages](#management) modular {{< glossary_tooltip term_id="resource" text="resources" >}} configured on your robot the same way it manages resources that are already built into the Robot Development Kit [(RDK)](/internals/rdk/).
```
To make clear that "like" doesn't mean "such as"

sguequierre replied:
Mostly just as the sentence has been up for a while and isn't new

JessamyT commented on docs/appendix/glossary/model-namespace-triplet.md:
```diff
@@ -0,0 +1,17 @@
+aka:
+---
+
+A model is a specific implementation of a resource that implements (speaks) its API.
+Models allow you to control different versions of resource types with a consistent interface.
+
+Models of a {{< glossary_tooltip term_id="resource" text="resource" >}} are uniquely namespaced as colon-delimited-triplets in the form of `namespace:family:name`.
```
We already have a glossary entry for "model" so I think this one can be shortened for triplet specificity.
```suggestion
{{< glossary_tooltip term_id="model" text="Models" >}} are uniquely namespaced as colon-delimited-triplets in the form of `namespace:family:name`.
```

sguequierre replied:
Awesome, thanks!

JessamyT commented on docs/extend/modular-resources/_index.md:
```diff
@@ -1,5 +1,5 @@
 ---
-title: "Create custom components and services"
+title: "Integrate Modular Resources into your robot"
```
```suggestion
title: "Integrate Modular Resources into your Robot"
```

JessamyT commented on docs/extend/modular-resources/_index.md:
```diff
@@ -10,679 +10,26 @@ aliases:
 
-The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}}, ([components](/components/), and [services](/services/)) into any robot running on Viam.
+The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}} into any robot running on Viam.
+The {{< glossary_tooltip term_id="rdk" text="RDK">}} [manages](/extend/modular-resources/key-concepts/) these modular resources like the resource types and models that are built-in to the kit.
```
```suggestion
The {{< glossary_tooltip term_id="rdk" text="RDK">}} [manages](/extend/modular-resources/key-concepts/) modular resources and built-in resources in the same way.
```
or can we avoid talking about RDK?
```suggestion
`viam-server` [manages](/extend/modular-resources/key-concepts/) modular resources and built-in resources in the same way.
```
(add link on viam-server maybe--and maybe we should add viam-server to the glossary but maybe that's for a different ticket)

npentrel replied:
the latter I think

JessamyT commented on docs/extend/modular-resources/_index.md:
```diff
@@ -10,679 +10,26 @@ aliases:
+Then, follow one of these [example tutorials](/extend/modular-resources/examples/) to add previously built modules and modular resources to your robot.
 
-### APIs
+If you want to create your own module providing custom resource models or new resource types as modular resources, follow [these instructions](/extend/modular-resources/create/) instead.
```
```suggestion
If you want to create your own module providing a new resource model or a new resource type as a modular resource, follow [these instructions](/extend/modular-resources/create/).
```

JessamyT commented on docs/extend/modular-resources/_index.md:
```diff
@@ -10,679 +10,26 @@ aliases:
+To get started adding custom resources to your robot, learn the [key concepts](/extend/modular-resources/key-concepts/) behind Viam's resource APIs that make the module system possible.
 
-## Key concepts
+Then, follow one of these [example tutorials](/extend/modular-resources/examples/) to add previously built modules and modular resources to your robot.
```
But some of these examples show creating new models? Perhaps it would make more sense to point them to the extend/modular-resources/configure/ page instead? And if necessary, that page could have links to specifically relevant examples.

sguequierre replied:
What do you mean? It's the rplidar and odrive, which I think are good simple pages to show the configuration process for pre-built modules. They do offer new models but that's in the module so I don't catch your drift

JessamyT replied:
Right now, the page looks like this: https://docs.viam.com/extend/modular-resources/examples/ It is true that those pre-built modules are contained in the examples folder, but we also have links to tutorials there that create new modules.

I think you should do this:
```suggestion
To add a previously built module to your robot, [configure a modular resource](/extend/modular-resources/configure/].
```

Then, if you want, links to relevant examples can be added to the bottom of the configure page.

Also, somewhat less importantly, this suggestion changes the order of words such that people know what they're doing before clicking the link (versus telling them to click the link before we actually know whether they want to use a prebuilt module or to make a new one).

JessamyT commented on docs/extend/modular-resources/_index.md:
```diff
@@ -10,679 +10,26 @@ aliases:
 
-### APIs
+If you want to create your own module providing custom resource models or new resource types as modular resources, follow [these instructions](/extend/modular-resources/create/) instead.
+After that, follow [these steps](/extend/modular-resources/configure/) to configure the module and modular resource.
```
```suggestion
Then, follow [these steps](/extend/modular-resources/configure/) to configure the module and modular resource.
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,85 @@
+
+### APIs
+
+Every Viam {{< glossary_tooltip term_id="resource" text="resource" >}} exposes an [Application Programming Interface (API)](https://en.wikipedia.org/wiki/API).
```
```suggestion
Every Viam {{< glossary_tooltip term_id="resource" text="resource" >}} type exposes an [application programming interface (API)](https://en.wikipedia.org/wiki/API).
```

sguequierre replied:
Same as above

JessamyT commented on docs/appendix/glossary/api-namespace-triplet.md:
```diff
@@ -0,0 +1,16 @@
+aka:
+---
+
+Every Viam {{< glossary_tooltip term_id="resource" text="resource" >}} exposes an [Application Programming Interface (API)](https://en.wikipedia.org/wiki/API) to describe how you can interact with that resource.
```
Really, we should add "type" to the glossary and put this in there, and then have the rest of this entry here. Not necessary right now unless you agree and feel so inclined.

sguequierre replied:
hmmm-- this is a good catch but now I'm unsure too. @mcvella wrote this section initially so this is not new material, I think it's worthwhile checking in with him-- would both of you consider this change to be more accurate? I guess I'm confused because the models are subtypes, but each subtype has their own resource API wrapper for the SDK, and each one is "exposing" that for communication? However, it is true that the APIs are defined by resource type (we don't use that phrasing a lot elsewhere to refer to types of components and services but that could also be an improvement)

npentrel replied:
agreed with adding type for clarification. Otherwise it does sound a big like a specific resource - like an instantiation of an arm. But I think this means every class/type of resource - like every subtype of arm would expose the same api. So every type sounds right to me.

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+
+The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}}, [components](/components/), and [services](/services/), into any robot running on Viam.
+
+A common use-case for modular resources is to create a new model using an existing Viam API.
```
```suggestion
A common use case for modular resources is to create a new model that implements an existing Viam API.
```
"use case" is correct. The other part is [opt].

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+
+### Code a new resource model
+
+{{% alert title="Note" color="note" %}}
```
I don't think this needs to be a note; it is just body text I think. Also, we're moving away from "note" to "important" or "stability notice" (and you can still use "tip" or "info" in green as well.

sguequierre replied:
```suggestion
```

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+
+{{% /alert %}}
+
+[Go Client SDK](https://pkg.go.dev/go.viam.com/rdk):
```
I think the format of this section is a bit difficult to parse visually because our expanders have bold text, whereas the surrounding "headings" aren't headings so the Python vs Go and the file names get lost visually. Maybe try using tabs like with the existing doc, and/or maybe put in horizontal lines to break up the sections? Or use headings or something? Something to emphasize the desired hierarchy/sections instead of the expanders being the biggest boldest thing.

sguequierre replied:
I agree. I tried to put the code in tabs but the expanders don't work inside of tabs and the code is chunky. Unsure @JessamyT or @npentrel do you have any ideas on this? maybe the lines?

JessamyT replied:
TBH I found the old format with the tabs and no expanders easier to follow/look at. Yeah I'd play with the "---" line separators perhaps. Maybe I'll mess around locally and make a suggestion if I come up with anything I actually like.

JessamyT replied:
Ooh another thing: since we do Python as the front tab, then Go, on most other pages, maybe do that here?

JessamyT replied:
Okay so, if you want to stick with this general organizational structure (as opposed to the way it was before) but use tabs, you could use tabs with the "details" tag like https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab and those do indeed nest within tabs.

JessamyT replied:
Also, I'm thinking you don't really need the link to the Go or Python SDKs on these "headings"

JessamyT replied:
With details tags, it looks like this: 
![image](https://github.com/viamrobotics/docs/assets/75634662/56fdae9d-dc13-4d1d-b723-3c450939ec36)
Here's a gist incorporating all my feedback on this section (does not include executable and next steps):
https://gist.github.com/JessamyT/514cd1da1694ade1bdff0a96a57bcda0

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+{{% /tab %}}
+{{% /tabs %}}
+
+### Code a new resource model
```
I like how you've created headings that match the steps described in the overview just above. However, what do people do for step one of define a new type of resource? Is there a way to point them to how to do that, if it's not covered right here?

JessamyT replied:
Oh, 762? Okay fair enough!

JessamyT commented on docs/extend/modular-resources/configure.md:
```diff
@@ -0,0 +1,145 @@
+2. [Add a **module**](#configure-your-module) referencing this executable to the configuration of your robot.
+3. [Add a new component or service](#configure-your-modular-resource) referencing the custom resource provided by the configured **module** to the configuration of your robot.
+
+{{% alert title="Modules vs. modular resources" color="tip" %}}
```
This tip may belong on the "key concepts" page

sguequierre replied:
It's covered there as well, I just thought it helps in here as it's so confusing

andf-viam replied:
I generally continue to be confused, so +1 from me to retain!

JessamyT replied:
Okay fine to retain here but where is it covered on the key concepts page? I don't see it covered there.

sguequierre replied:
Oops I thought I put that alert in there too but I didn't keep it! adding

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+
+Follow [these instructions](/extend/modular-resources/configure/) to add your custom resource to your robot.
+
+Make your [executable accessible](/extend/modular-resources/configure/#make-sure-viam-server-can-access-your-executable), [configure your new module](/extend/modular-resources/configure/#configure-your-module), and [configure your modular resource](/extend/modular-resources/configure/#configure-your-modular-resource).
```
```suggestion
```
This page just had them do all this so I don't think it makes sense to put this in next steps.

sguequierre replied:
I think it helps to provide an overview of what they haven't done yet? After completing the create page they haven't done that

JessamyT replied:
Ahh sorry I misread. You're totally right. However, I still think it's unnecessary to list this all out with links since if they click the link above it takes them to all of this. Will make a different suggestion.

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+Follow [these instructions](/extend/modular-resources/configure/) to add your custom resource to your robot.
+
+Make your [executable accessible](/extend/modular-resources/configure/#make-sure-viam-server-can-access-your-executable), [configure your new module](/extend/modular-resources/configure/#configure-your-module), and [configure your modular resource](/extend/modular-resources/configure/#configure-your-modular-resource).
+
+After configuring a modular resource, test it with the [Control tab](/manage/fleet/#remote-control) and program it with the [Viam SDKs](/program/apis/).
```
```suggestion
```
I think one next step (configure) makes sense here, and for clarity it's best to just leave it at that. Maybe add this step to the bottom of the configure page though!! I think that'd be great!

sguequierre replied:
Same as above

JessamyT replied:
I'm saying you should move this line to the bottom of the configure instructions since it's not relevant to people here yet, until after they configure. I think of the "Next steps" as being an actionable place where we guide them to exactly what to do next, and I don't want them to go try and test stuff when they can't yet.

JessamyT commented on docs/extend/modular-resources/configure.md:
````diff
@@ -0,0 +1,145 @@
+```
+
+{{% /tab %}}
+{{% /tabs %}}
````
```suggestion
{{% /tabs %}}

## Next steps

Now that you've configured a modular resource, test it with the [Control tab](/manage/fleet/#remote-control) and program it with the [Viam SDKs](/program/apis/).
```
[opt]

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+You need to ensure any dependencies for your module (including the Viam SDK) are installed, as well.
+Your executable will be run by `viam-server` as root, so dependencies need to be available to the root user.
+
+## Next Steps
```
```suggestion
## Next steps
```
[nit] The other headings are sentence case

JessamyT reviewed (changes requested):
Nice--I like how you've organized a lot of this so it'll be easier to link people to parts that are directly relevant to them.

sguequierre commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,515 @@
+
+The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
+
+{{% /alert %}}
```
```suggestion
```

sguequierre commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,513 @@
+
+The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
+
+
```
```suggestion
```

sguequierre commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,513 @@
+
+### Code a new resource model
+
+
```
```suggestion
```

npentrel commented on docs/appendix/glossary/api-namespace-triplet.md:
```diff
@@ -0,0 +1,16 @@
+
+Every Viam {{< glossary_tooltip term_id="resource" text="resource" >}} exposes an [Application Programming Interface (API)](https://en.wikipedia.org/wiki/API) to describe how you can interact with that resource.
+
+Each Viam resource's API is uniquely namespaced as a colon-delimited-triplet in the form of `namespace:type:subtype`.
```
can we avoid passive voice to make it clearer for the user what being namespaced means (here and in other instances)

npentrel replied:
I might be wrong here and this is how people generally speak about it - In cases like this I generally google a phrase like "namespaced component" or "namespaced object" and see how often it comes up. In this case there a few thousand instances. "an object in the namespace" or defined by the namespace is I think what I would say? @mcvella do you have opinions on this?

sguequierre replied:
```suggestion
 These APIs are organized by colon-delimited-triplet identifiers, in the form of `namespace:type:subtype`.
 
 The `namespace` for built-in Viam resources is `rdk`, while the `type` is `component` or `service`.
 `subtype` refers to a specific component or service, like a `camera` or `vision`.

One subtype can have various models, custom or built-in, but they all must conform to the subtype's API definition.
This requirement ensures that when a resource of that model is deployed, you can [interface with it](/program/) using the same [client API methods](/program/apis/) you would when programming resources of the same subtype with a different model.
```
passive voice suggestion at top
extra sentences are likely overboard but could be helpful info?

sguequierre replied:
@npentrel @mcvella

mcvella replied:
I like Sierra's suggested change above, its verbose but in a helpful way

npentrel replied:
looks good to me

npentrel commented on docs/appendix/glossary/api-namespace-triplet.md:
```diff
@@ -0,0 +1,16 @@
+
+For example:
+
+- The API of built-in component [camera](/components/camera/) is `rdk:component:camera`, which exposes methods such as `GetImage()`.
+- The API of built-in service [vision](/services/vision/) is `rdk:service:vision`, which exposes methods such as `GetDetectionsFromCamera()`.
```
```suggestion
- The API of the built-in component [camera](/components/camera/) is `rdk:component:camera`, which exposes methods such as `GetImage()`.
- The API of the built-in service [vision](/services/vision/) is `rdk:service:vision`, which exposes methods such as `GetDetectionsFromCamera()`.
```

npentrel commented on docs/appendix/glossary/model-namespace-triplet.md:
```diff
@@ -0,0 +1,14 @@
+
+For example:
+
+- The `rdk:builtin:gpio` model of the `rdk:component:motor` API provides RDK support for [GPIO-controlled DC motors](/components/motor/gpio/).
+- The `rdk:builtin:DMC4000` model of the same `rdk:component:motor` API provides RDK support for the [DMC4000](/components/motor/dmc4000/) motor.
```
we probably want to take out RDK here and implementation may be a better word than support here

JessamyT replied:
Ooh this is a tricky RDK usage. Since "rdk" is part of the triplet, maybe it actually makes sense to leave it in here for context? Also I think "support" makes sense to me here because I think of it as, the code is the implementation that supports the hardware. In other words, `rdk:builtin:gpio` is an implementation of the `rdk:component:motor` API, and that implementation of that API supports a given motor.

Here's a suggestion if "RDK" comes out:
```suggestion
- The `rdk:builtin:gpio` model of the `rdk:component:motor` API supports [GPIO-controlled DC motors](/components/motor/gpio/).
- The `rdk:builtin:DMC4000` model of the same `rdk:component:motor` supports the [DMC4000](/components/motor/dmc4000/) motor controller.
```
Maybe the following would use the word "implement" correctly?
```suggestion
- The `rdk:builtin:gpio` model implements the `rdk:component:motor` API to support [GPIO-controlled DC motors](/components/motor/gpio/).
- The `rdk:builtin:DMC4000` model of the same `rdk:component:motor` supports the [DMC4000](/components/motor/dmc4000/) motor controller.
```

sguequierre replied:
I think this suggestion looks good but @mcvella as author does this change make sense to you?

mcvella replied:
how about saying "provides built-in" support?  This way we don't have to say RDK but imply that it is part of RDK/viam-server?  Otherwise its fine as-is, since some of these things might become modular in the future anyway!

sguequierre replied:
```suggestion
- The `rdk:builtin:gpio` model implements the `rdk:component:motor` API for [GPIO-controlled DC motors](/components/motor/gpio/).
- The `rdk:builtin:DMC4000` model implements the `rdk:component:motor` API for [DMC4000](/components/motor/dmc4000/) motor controllers.
```

npentrel reviewed (commented):
Good improvements I think - left a few comments :)

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+
+## Next steps
+
+Follow [these instructions](/extend/modular-resources/configure/) to add your custom resource to your robot.
+
+Make your [executable accessible](/extend/modular-resources/configure/#make-sure-viam-server-can-access-your-executable), [configure your new module](/extend/modular-resources/configure/#configure-your-module), and [configure your modular resource](/extend/modular-resources/configure/#configure-your-modular-resource).
```
```suggestion
Follow [these configuration instructions](/extend/modular-resources/configure/) to add your custom resource to your robot.
```
I think it's best to leave it at that so people don't feel the need to click all the links separately or fear they're missing something, when one link has everything they need to do.

JessamyT reviewed (commented):
Looks like there's also still some outstanding comments from the last round.

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+
+The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/extend/modular-resources/key-concepts/#models) as a new model, `"mybase"`:
+
+The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
```
```suggestion
The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) are adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
```

JessamyT replied:
Oh man my suggestion is grammatically incorrect sorry!

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+
+<file>mybase.go</file>
+
+{{%expand "Click to view example code from a Go module implementing a custom model of the base component built-in resource, mybase, and registering the new model and API" %}}
```
```suggestion
{{%expand "Click to view example code from a Go module implementing a custom model of the base component built-in resource, mybase, and registering the new model" %}}
```
I think this could confuse people and make people think they're creating a new API instead of implementing an existing one

sguequierre replied:
Maybe should add the helpers clarification?

sguequierre replied:
like 
```suggestion
{{%expand "Click to view example code from a Go module implementing a custom model of the base component built-in resource, mybase, and registering the new model and API helpers with the client SDK" %}}
```
Struggling as it's a little long but lmk what you think. can incorporate this feedback either way when I get time to rework the whole section re feedback too

JessamyT replied:
I'm not sure what the "helpers" adds? I think shorten it as much as possible, and if you have to you can include other details inside the expander or in the body text. See also my formatting suggestion though. And, if you come up with an entirely different format, this could become moot anyway? But if not, maybe:
```suggestion
{{%expand "Click to view example code from a Go module implementing and registering a custom model of the base component" %}}
```

sguequierre replied:
done, and added details!

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+
+<file>_\_init__.py</file>
+
+{{%expand "Click to view example code from a Python module registering the mybase custom model and API" %}}
```
```suggestion
{{%expand "Click to view example code from a Python module registering the mybase custom model" %}}
```
See above

sguequierre replied:
changed to details and took that out of the clicker

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+
+<file>main.go</file>
+
+{{%expand "Click to view example code from a Go module's entry point file initializing the mybase custom model and API from the registry" %}}
```
```suggestion
{{%expand "Click to view example code from a Go module's entry point file initializing the mybase custom model from the registry" %}}
```
As above--are they actually registering an API or just a model that uses that API?

sguequierre replied:
registering helpers w SDK for both I think, it's confusing

JessamyT replied:
Hey for the sake of clarity etc. instead of editing my suggestions themselves, mind making a new suggestion on the thread and copy pasting from mine if you want?

What is an API helper? I'm still thinking "API" should come out of here

JessamyT replied:
But I do like taking "from the registry" off too since it's long

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+
+<file>main.py</file>
+
+{{%expand "Click to view example code from a Python module's entry point file initializing the mybase custom model and API from the registry" %}}
```
```suggestion
{{%expand "Click to view example code from a Python module's entry point file initializing the mybase custom model from the registry" %}}
```
see above

JessamyT reviewed (commented):
additional comments

JessamyT commented on docs/extend/modular-resources/create.md:
````diff
@@ -0,0 +1,511 @@
+
+To make this shell script executable, run the following command in your terminal:
+
+``` shell
````
```suggestion
```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
```
To make it have a black background and more clearly be a terminal command

sguequierre replied:
done

JessamyT commented on docs/extend/modular-resources/create.md:
````diff
@@ -0,0 +1,511 @@
+```
+
+{{% /tab %}}
+{{% /tabs %}}
````
```suggestion
{{< /tabs >}}
```
[nit] Just because the opening one is like this, so makes sense to match them I think

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+To create your own modular resource, code a module in Go or Python using the module support libraries provided by [Viam's SDKs](/program/apis/) that implements at least one new model or type of {{< glossary_tooltip term_id="resource" text="resource" >}}:
+
+{{< tabs >}}
+{{% tab name="Define a New Model of a Built-In Resource Type" %}}
+
```
```suggestion
{{% tab name="New Model" %}}
Define a new model of a built-in resource type:

```
[opt] Realized these tab names are so long that they show up stacked oddly on many screens.

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,511 @@
+3. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable that can receive a socket argument from Viam, open the socket, and start the module at the entry point.
+
+{{% /tab %}}
+{{% tab name="Define a New Type of Resource" %}}
+
```
```suggestion
{{% tab name="New Type" %}}
Define a new type of resource:

```
see above, opt

JessamyT commented:
Another option for reorganizing the tabs/expanders/details/steps on the create page: Make sub-pages of Create for new model and for new type. Right now, there's not a great place for the additional info on defining a new type to go (https://viam.atlassian.net/browse/DOCS-762), but if there were separate sub-pages it might make things easier to format and follow, and we could either link to them as necessary or use snippets if there is some re-used content between the two.

npentrel commented:
going to hold off with a bigger review since there's still lots being discussed - just two questions:
- does an expander inside a tab not work?
- are those code files present in the sdk docs or as templates somewhere? I'm worried they may grow outdated without us knowing so linking to the source files if they exist may be a good idea.

sguequierre commented:
> going to hold off with a bigger review since there's still lots being discussed - just two questions:
> 
> * does an expander inside a tab not work?
> * are those code files present in the sdk docs or as templates somewhere? I'm worried they may grow outdated without us knowing so linking to the source files if they exist may be a good idea.

answers! 

Expanders inside an (%) tab wind up looking like this: 
<img width="755" alt="image" src="https://github.com/viamrobotics/docs/assets/20797913/c992632f-d8fd-4937-b0d2-7e78dc2b12a9">

The go code is heavily modified from a linked source repo, the python __init__.py is slightly modified from the same eng. demo repo as well as think-- but the python code i pretty much wrote myself 
I think leaving the link as is might be the easiest solution. I don't think I could dump my python template into the eng demos repo but I could create a place in Viam labs for it, that might be overkill though

sguequierre commented:
@npentrel @JessamyT I implemented a bunch of suggestions including making <details> sections instead of using expand for the files-- I think it is ready for re-review now but feel free to let me know directly if I missed any of your feedback

mcvella commented on docs/extend/modular-resources/configure.md:
```diff
@@ -0,0 +1,149 @@
+If you want to create your own modular resource, you must first follow [these steps](/extend/modular-resources/create/) to code your own module and generate an executable.
+If you are using a pre-built modular resource, make sure you install the module and determine the filename of [the module's executable](/extend/modular-resources/create/#compile-the-module-into-an-executable).
+
+Follow these steps to configure the module and modular resource:
```
Shouldn't we say resources?  Since multiple can be provided by a single module?

JessamyT replied:
I think that this is saying "hey you are adding a modular resource here's how you do that one task," rather than saying "here's how modular resources are configured in general," so I think the singular works well

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) are adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
+
+{{< tabs name="Sample SDK Code">}}
+{{% tab name="Go"%}}
```
Is there a reason for showing Go before Python on all of these, contrary to what we do in the rest of the docs?

sguequierre replied:
No just switched up in my head, fixed!

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+{{% tab name="Go"%}}
+
+<file>main.go</file> is the demo Go module's entry point file.
+When executed, it initializes the mybase custom model and API helper functions from the registry.
```
```suggestion
When executed, it initializes the `mybase` custom model.
```
I think the way the custom model uses the API *is* the custom model, so I think this can be shortened.

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+{{% tab name="Python"%}}
+
+<file>main.py</file> is the demo Python module's entry point file.
+When executed, it initializes the mybase custom model and API helper functions from the registry.
```
```suggestion
When executed, it initializes the `mybase` custom model.
```
see above; I think this is redundant?

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+{{< tabs name="Sample SDK Main Program Code">}}
+{{% tab name="Go"%}}
+
+<file>main.go</file> is the demo Go module's entry point file.
```
```suggestion
<file>main.go</file> is the Go module's entry point file.
```
I think reads better

sguequierre replied:
ok!

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -59,15 +59,17 @@ The following example module registers a modular resource implementing Viam's bu
 
 The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) are adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
```
```suggestion
The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
```
sorry I realize I accidentally suggested the opposite earlier

sguequierre replied:
fixed

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+
+### Models
+
+{{% alert title="Modules vs. modular resources" color="tip" %}}
```
I think actually move this tip up to just above the API heading. This section on models isn't about modules vs modular resources so I think it fits better up there!

sguequierre replied:
done!

JessamyT commented on docs/extend/modular-resources/configure.md:
```diff
@@ -0,0 +1,149 @@
+
+### Configure your modular resource
+
+Once you have configured a module as part of your robot configuration, you can add any number of the resources that module makes available to your robot by adding new components or services configured with your modular resources' new {{< glossary_tooltip term_id="api-namespace-triplet" text="subtype" >}} or [model](/extend/modular-resources/key-concepts/#models).
```
```suggestion
Once you have configured a module as part of your robot configuration, you can add any number of the resources that module makes available to your robot by adding new components or services configured with your modular resources' new subtype or [model](/extend/modular-resources/key-concepts/#models).
```
I think the "namespace:type:subtype" tooltip isn't helpful here since we're not talking about namespacing right here.

JessamyT commented on docs/extend/modular-resources/examples/_index.md:
```diff
@@ -3,7 +3,7 @@ title: "Modular Resource Configuration Examples"
 linkTitle: "Examples"
 childTitleEndOverwrite: "Modular Resource Example"
```
Not part of diff but what is this for?

sguequierre replied:
not sure, I don't think I added

JessamyT commented on docs/extend/modular-resources/examples/_index.md:
I think this page should be changed to either have a different title, or only include config-specific examples. I'd advocate the former, since we want all the modular resource examples, so maybe change title to "Modular Resource Examples" and then there could be a subsection (if you want) of config ones vs creation ones, or not. But def change title.

sguequierre replied:
I think this section is ok as is but that would be an improvement!

JessamyT replied:
Yeah I realize it's not part of the diff but this whole section is being refactored in this PR so I thought it's a good time to look at all of it. If you think it's an improvement, it's easy to loop that one line change into this PR! Don't have to though.

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+`viam-server` [manages](#management) modular {{< glossary_tooltip term_id="resource" text="resources" >}} configured on your robot the same way it manages resources that are already built into the Robot Development Kit [(RDK)](/internals/rdk/).
+Two key concepts exist across all Viam resources, both built-in and modular, that make this flexible management possible: uniquely namespaced resource [*APIs*](#apis) and [*models*](#models).
+
+If you create your [own module](/extend/modular-resources/create/), you must register any new APIs and models you define in your model with Viam's model registry in the appropriate namespaces to [configure](/extend/modular-resources/configure/) the modular resource on your robot.
```
```suggestion
If you create your [own module](../create/), you must register any new APIs and models you define in your model with Viam's model registry in the appropriate namespaces to [configure](../configure/) the modular resource on your robot.
```

JessamyT replied:
Okay IDK what the latest policy is on this and I don't feel strongly!

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+no_list: true
+---
+
+The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}}, [components](/components/), and [services](/services/), into any robot running on Viam.
```
```suggestion
The Viam module system allows you to integrate custom [components](/components/) and [services](/services/) into any robot running on Viam.
```
Or leave resources and take out components and services, or do 
```suggestion
The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}} ([components](/components/) and [services](/services/)) into any robot running on Viam.
```

JessamyT commented on docs/extend/modular-resources/create.md:
````diff
@@ -0,0 +1,528 @@
+        return self.left.is_powered(extra=extra, timeout=timeout)[0] or self.right.is_powered(extra=extra, timeout=timeout)[0]
+```
+
+<br>
+</details>
+
````
```suggestion

</details>
<br>

```

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+<br>
+</details>
+
+<file>__init__.py</file> registers the mybase custom model and API helper functions with the SDK.
```
right now in the rendering, init is showing up bold instead of with the underscores.

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+{{% tabs %}}
+{{% tab name="Python" %}}
+
+If using the Python SDK, one option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.
```
```suggestion
Create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.
```
They're already on the Python tab so you can take that part out. What is the other option if not this option? Either say what it is or take that out I think.

sguequierre replied:
I was working on this section refactor with james quite a bit recently, the diff would just be some words if I hadn't moved the text from a different page.  In my understanding the wording is intentional bc there are a lot of different ways you can set the module executable. This is what James suggested as simplest use case for each sdk

JessamyT commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,528 @@
+{{% /tab %}}
+{{% tab name="Go" %}}
+
+If using the Go SDK, use Go to [compile](https://pkg.go.dev/cmd/go#hdr-Compile_packages_and_dependencies) and obtain a single executable for your module:
```
```suggestion
Use Go to [compile](https://pkg.go.dev/cmd/go#hdr-Compile_packages_and_dependencies) and obtain a single executable for your module:
```

JessamyT commented on docs/extend/modular-resources/configure.md:
```diff
@@ -0,0 +1,149 @@
+
+For example, if you are running `viam-server` on an Raspberry Pi, you'll need to save the module on the Pi's filesystem.
+
+Obtain the real (absolute) path to the executable file on your computer/[board's](/components/board/) filesystem by running the following command in your terminal:
```
```suggestion
Obtain the real (absolute) path to the executable file on your computer's filesystem by running the following command in your terminal:
```

JessamyT commented on docs/extend/modular-resources/configure.md:
```diff
@@ -0,0 +1,149 @@
+| `type` | string | **Required** | The subtype of the [API](/extend/modular-resources/key-concepts/#apis) (the third part of the {{< glossary_tooltip term_id="api-namespace-triplet" text="API namespace triplet">}}). |
+| `name` | string | **Required** | What you want to name this instance of your modular resource. |
+| `model` | string | **Required** | The full {{< glossary_tooltip term_id="model-namespace-triplet" text="model namespace triplet">}} of the modular resource's [model](/extend/modular-resources/key-concepts/#models). |
+| `depends_on` | array | Optional | The `name` of components you want to confirm are available on your robot alongside your modular resource. Usually a [board](/components/board/). |
```
```suggestion
| `depends_on` | array | Optional | The `name` of components you want to confirm are available on your robot alongside your modular resource. Often a [board](/components/board/). Unnecessary if you coded [implicit dependencies](../#dependency-management). |
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+For example:
+
+Some DC motors use just [GPIO](/components/board/), while other DC motors use serial protocols like [SPI bus](/components/board/#spis).
+Regardless, you can power any motor model that implements the *rdk:component:motor* API with the `SetPower()` method.
```
```suggestion
Regardless, you can power any motor model that implements the `rdk:component:motor` API with the `SetPower()` method.
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+
+If you create your [own module](/extend/modular-resources/create/), you must register any new APIs and models you define in your model with Viam's model registry in the appropriate namespaces to [configure](/extend/modular-resources/configure/) the modular resource on your robot.
+
+### APIs
```
```suggestion
## APIs
```
All of these headings can go up one

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+Each API is described through [protocol buffers](https://developers.google.com/protocol-buffers).
+Viam SDKs [expose these APIs](/internals/robot-to-robot-comms/).
+
+#### Namespace
```
```suggestion
### Namespace
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+You can see built-in Viam resource APIs in the [Viam GitHub](https://github.com/viamrobotics/api).
+{{% /alert %}}
+
+### Models
```
```suggestion
## Models
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+Some DC motors use just [GPIO](/components/board/), while other DC motors use serial protocols like [SPI bus](/components/board/#spis).
+Regardless, you can power any motor model that implements the *rdk:component:motor* API with the `SetPower()` method.
+
+#### Namespace
```
```suggestion
### Namespace
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+- The `rdk:builtin:gpio` model of the `rdk:component:motor` API provides RDK support for [GPIO-controlled DC motors](/components/motor/gpio/).
+- The `rdk:builtin:DMC4000` model of the same `rdk:component:motor` API provides RDK support for the [DMC4000](/components/motor/dmc4000/) motor.
+
+### Management
```
```suggestion
## Management
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+
+The Robot Development Kit (RDK) `viam-server` provides automatically manages modular resources to function like built-in resources:
+
+#### Dependency Management
```
```suggestion
### Dependency Management
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+Modular resources may depend on other built-in resources or other modular resources, and vice versa.
+The Viam RDK handles dependency management.
+
+#### Start-up
```
```suggestion
### Start-up
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+
+The RDK ensures that any configured modules are loaded automatically on start-up, and that configured modular resource instances are started alongside configured built-in resource instances.
+
+#### Reconfiguration
```
```suggestion
### Reconfiguration
```

sguequierre replied:
Wait, why do you want to change all of these headings to be visible in the sidebar? Is this necessary?

JessamyT replied:
There is no H2 above them all, so why make them so small? 

Additionally, there's less of a different between H3 and H4 than there is between H2 and H3 (at least to my eye), so it is easier to parse when it's that way, and here, why not??

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+When you change the configuration of a Viam robot, the behavior of modular resource instances versus built-in resource instances is equivalent.
+This means you can add, modify, and remove a modular resource instance from a running robot as normal.
+
+#### Data management
```
```suggestion
### Data management
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+Data capture for individual components is supported on [certain component subtypes](/services/data/configure-data-capture/#configure-data-capture-for-individual-components).
+If your modular resource is a model of one of these subtypes, you can configure data capture on it just as you would on a built-in resource.
+
+#### Shutdown
```
```suggestion
### Shutdown
```

JessamyT commented on docs/extend/modular-resources/key-concepts.md:
```diff
@@ -0,0 +1,91 @@
+
+During robot shutdown, the RDK handles modular resource instances similarly to built-in resource instances - it signals them for shutdown in topological (dependency) order.
+
+#### Modular resources as remotes
```
```suggestion
### Modular resources as remotes
```

JessamyT commented on docs/tutorials/custom/controlling-an-intermode-rover-canbus.md:
```diff
@@ -95,7 +95,7 @@ Check out this [GitHub repository](https://github.com/viam-labs/tutorial-intermo
 ## A modular resource for the Intermode base
 
-Viam includes [APIs](/extend/modular-resources/#apis) for common component types within `viam-server`.
+Viam includes [APIs](/extend/modular-resources/key-concepts/#apis) for common component types within `viam-server`.
```
```suggestion
Viam includes [APIs](/program/apis/) for common component types within `viam-server`.
```
I think this would make more sense here.

sguequierre commented on docs/extend/modular-resources/_index.md:
```diff
@@ -10,686 +10,30 @@ aliases:
+- To create your own module and modular resources, follow [these instructions](/extend/modular-resources/create/).
 
-### APIs
+- To download modules and add modular resources that other users have built to your robot, follow one of Viam's [module setup guides](/extend/modular-resources/examples/).
```
Hmmm IMO it's not essentially the same as configure though-- eventually those module docs that are up on the site will probably mostly be the module descriptions and install instructions + what specific config attributes

From my view this bullet is helpful for people that want to browse options for pre-built modules and are not set on questing the create

sguequierre commented:
@JessamyT I think this is ready for re-review!

JessamyT reviewed (commented):
There are still a couple of unaddressed comments on /create.md

sguequierre commented:
> There are still a couple of unaddressed comments on /create.md

see my comments there about the wording ambiguity, unless they are mandatory I think that section is good as is!

sguequierre commented on docs/extend/modular-resources/create.md:
```diff
@@ -0,0 +1,527 @@
+{{% tabs %}}
+{{% tab name="Python" %}}
+
+If using the Python SDK, one option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.
```
```suggestion
One option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.
```

sguequierre replied:
@JessamyT with this suggestion it might address your concern but keep the wording ambiguous?

JessamyT replied:
@sguequierre The goal is to not be ambiguous :D But okay I guess there's the mention of flexibility just above the tabs. To make it less ambiguous you could say "The simplest option is to create and save..." instead of "One option..." but not going to block on this

JessamyT reviewed (approved):
% recommend changing "Modular Resource Configuration Examples" to "Modular Resource Examples" as discussed
